### PR TITLE
change health-check port to listen to node port addresses

### LINKED ISSUE #103184: healthCheckNodePort in service object cannot listen on specific ip address
#### What happened:
healthCheckNodePort in service object listen on 0.0.0.0, and cannot specify ip address.

#### What you expected to happen:
healthCheckNodePort in service object can listen on specific ip address **in some security scene, such as multi network interface**

#### How to reproduce it (as minimally and precisely as possible):
create a service object with filling healthCheckNodePort field

#### Anything else we need to know?:
nodePort already support to listen on specific ip address through "--nodeport-addresses" option in kube-proxy, does community plan to add same enhancement in healthCheckNodePort ?

#### Environment:
- Kubernetes version (use `kubectl version`): 1.19
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by khenidak**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
kube-proxy health check ports used to listen to `:<port>` for each of the services. This is not needed and opens ports in addresses the cluster user may not have intended. The PR limits listening to all node address which are controlled by `--nodeport-addresses` flag.

#### Which issue(s) this PR fixes:
Fixes #103184

#### Special notes for your reviewer:
WIP don't review yet.

#### Does this PR introduce a user-facing change?
```release-note
kube-proxy health check ports used to listen to :<port> for each of the services. This is not needed and opens ports in addresses the cluster user may not have intended. The PR limits listening to all node address which are controlled by `--nodeport-addresses` flag. if no addresses are provided then we default to existing behavior by listening to :<port> for each service
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

/sig network
/hold

### CODE REVIEW

**Comment by aojea** [comment]
the unit test failing seems related
```
    healthcheck_test.go:154: expected 1 open port, got 0
        (sets.String) {
        }
    healthcheck_test.go:157: expected port :9376 to be open
        (sets.String) {
        }
panic: interface conversion: healthcheck.httpServer is nil, not *healthcheck.fakeHTTPServer [recovered]
	panic: interface conversion: healthcheck.httpServer is nil, not *healthcheck.fakeHTTPServer
```

**Comment by aojea** [comment]
I think this should be aligned to current logic
https://github.com/kubernetes/kubernetes/blob/eb729620c522753bc7ae61fc2c7b7ea19d4aad2f/pkg/proxy/iptables/proxier.go#L996-L1000

**Comment by khenidak** [comment]
Yes. I kept that one failing until i clean up the old code. The current commits are the change i want to make. But i am not happy with unit testing coverage so i am keeping the `wip` tag on until i figure out a way to cover selective addresses

**Comment by khenidak** [comment]
@danwinship @aojea -- Can you help reviewing this? (it is ready now)

**Comment by danwinship** [comment]
> This is not needed and opens ports in addresses the cluster user may not have indented

"intended" :slightly_smiling_face: 
(in the release note text)

File: pkg/proxy/healthcheck/healthcheck_test.go [hunk 1]
```diff
@@ -411,3 +415,52 @@ func testHealthzHandler(server httpServer, status int, t *testing.T) {
...
 }
+
+func TestServerWithSelectiveListeningAddress(t *testing.T) {
+	listener := newFakeListener()
+	httpFactory := newFakeHTTPServerFactory()
+
+	/*
+		limiting addresses to loop back. We don't want any cleverness here around getting IP for
+		machine nor testing ipv6 || ipv4. using loop back guarantees the test will work on any machine
+	*/
```
Comment (danwinship) [inline pkg/proxy/healthcheck/healthcheck_test.go:426]: we use //-style comments pretty much everywhere

File: pkg/proxy/healthcheck/service_health.go [hunk 2]
```diff
@@ -117,25 +141,75 @@ func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) err
...
+			return err
+		}
+
+		// start serving
+		go func(hcI *hcInstance, listener net.Listener, httpSrv httpServer) {
 			// Serve() will exit when the listener is closed.
-			klog.V(3).Infof("Starting goroutine for healthcheck %q on port %d", nsn.String(), svc.port)
-			if err := svc.server.Serve(svc.listener); err != nil {
-				klog.V(3).Infof("Healthcheck %q closed: %v", nsn.String(), err)
+			//klog.V(3).Infof("Starting goroutine for healthcheck %q on  %s", hcI.nsn.String(), listener.Addr().String())
```
Comment (danwinship) [inline pkg/proxy/healthcheck/service_health.go:184]: any reason you commented this out? Seems like it should either be uncommented or deleted

**Review by danwinship — COMMENTED** [review]
mostly looks good

File: pkg/proxy/healthcheck/healthcheck_test.go (hunk 1 above)
Comment (khenidak) [reply pkg/proxy/healthcheck/healthcheck_test.go:426]: ack

File: pkg/proxy/healthcheck/service_health.go (hunk 2 above)
Comment (khenidak) [reply pkg/proxy/healthcheck/service_health.go:184]: ack -- will added back.

**Comment by danwinship** [comment]
/lgtm
but not un-holding it since I'm not sure if there was some other reason for that

**Comment by khenidak** [comment]
The hold is to get as many eyes as possible on it. One can argue it is a behavior change so i am not sure how to deal with this one. There is no api change, just behavior change. Do we need anything more than release notes?

@thockin ?

File: pkg/proxy/healthcheck/service_health.go [hunk 3]
```diff
@@ -117,25 +141,75 @@ func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) err
...
+	var err error
+	var listener net.Listener
+
+	addresses := hcs.nodeAddresses.List()
+	hcI.listeners = make([]net.Listener, 0, len(addresses))
+	hcI.httpServers = make([]httpServer, 0, len(addresses))
+
+	// for each of the node addresses start listening and serving
+	for _, address := range addresses {
+		addr := fmt.Sprintf("%s:%d", address, hcI.port)
```
Comment (aojea) [inline pkg/proxy/healthcheck/service_health.go:169]: net.JoinHostPort

File: pkg/proxy/healthcheck/healthcheck_test.go [hunk 4]
```diff
@@ -411,3 +424,51 @@ func testHealthzHandler(server httpServer, status int, t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestServerWithSelectiveListeningAddress(t *testing.T) {
+	listener := newFakeListener()
+	httpFactory := newFakeHTTPServerFactory()
+
+	// limiting addresses to loop back. We don't want any cleverness here around getting IP for
+	// machine nor testing ipv6 || ipv4. using loop back guarantees the test will work on any machine
```
Comment (aojea) [inline pkg/proxy/healthcheck/healthcheck_test.go:433]: are not fake the listeners?

File: pkg/proxy/healthcheck/service_health.go [hunk 5]
```diff
@@ -96,11 +121,10 @@ func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) err
 		}
 
 		klog.V(2).Infof("Opening healthcheck %q on port %d", nsn.String(), port)
```
Comment (aojea) [inline pkg/proxy/healthcheck/service_health.go:123]: ... on port %d and node addresses %v

File: pkg/proxy/healthcheck/service_health.go [hunk 6]
```diff
@@ -117,25 +141,75 @@ func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) err
...
+				klog.V(3).Infof("Healthcheck %q closed: %v", hcI.nsn.String(), err)
 				return
 			}
-			klog.V(3).Infof("Healthcheck %q closed", nsn.String())
-		}(nsn, svc)
+			klog.V(3).Infof("Healthcheck %q on %s closed", hcI.nsn.String(), listener.Addr().String())
+		}(hcI, listener, httpSrv)
+
+		hcI.listeners = append(hcI.listeners, listener)
+		hcI.httpServers = append(hcI.httpServers, httpSrv)
```
Comment (aojea) [inline pkg/proxy/healthcheck/service_health.go:193]: if the httpSrv contains the listener, do we need to store both? 

can it be racy if we close a server and a listener for that server?

Comment (aojea) [reply pkg/proxy/healthcheck/service_health.go:193]: nvm, we only close the listeners

File: pkg/proxy/healthcheck/healthcheck_test.go (hunk 4 above)
Comment (khenidak) [reply pkg/proxy/healthcheck/healthcheck_test.go:433]: They are. They also have addresses handling logic that mimics the real thing.

Comment (aojea) [reply pkg/proxy/healthcheck/healthcheck_test.go:433]: Then we can use non loopback addresses, the comment is not correct , no?

**Review by thockin — COMMENTED** [review]
This seems OK to me, but I wonder if there are use-cases we're missing.  I guess we'll find out.  I expected to see `map[types.NamespacedName][]*hcInstance` but this works, too :)

/lgtm
/approve

**Pull request merged**